File: backend/app/services/entity_resolution.py
"""Entity resolution: background merge of duplicate nodes.

Strategy (decide early, per plan): normalized-name + alias overlap blocking,
then Jaro-Winkler-ish scoring. Merges are explicit (survivor + retired ids)
and logged to audit — never silent. This module implements the pure,
testable scoring; the Neo4j rewrite runs as an arq background job.
"""

import re
import unicodedata
from difflib import SequenceMatcher


def normalize(name: str) -> str:
    name = unicodedata.normalize("NFKD", name).encode("ascii", "ignore").decode()
    name = name.lower()
    name = re.sub(r"[.\-_,]", " ", name)
    name = re.sub(r"\s+", " ", name).strip()
    # Expand common initials pattern "n chandrasekaran" stays as-is; drop titles
    name = re.sub(r"^(mr|mrs|ms|dr|prof)\s+", "", name)
    return name


def _token_overlap(ta: set[str], tb: set[str]) -> float:
    """Overlap counting single-letter initials as prefix matches ('n' ~ 'natarajan')."""
    if not ta or not tb:
        return 0.0
    small, big = (ta, tb) if len(ta) <= len(tb) else (tb, ta)
    matched = 0
    for tok in small:
        if tok in big:
            matched += 1
        elif len(tok) == 1 and any(b.startswith(tok) for b in big):
            matched += 1
    return matched / max(len(ta), len(tb))


def name_score(a: str, b: str) -> float:
    na, nb = normalize(a), normalize(b)
    if not na or not nb:
        return 0.0
    if na == nb:
        return 1.0
    # token-set overlap boosts "N Chandrasekaran" vs "Natarajan Chandrasekaran"
    ta, tb = set(na.split()), set(nb.split())
    overlap = _token_overlap(ta, tb)
    seq = SequenceMatcher(None, na, nb).ratio()
    # last-token (family name) must match for a merge candidate
    if na.split()[-1] != nb.split()[-1]:
        return min(seq, 0.5)
    return max(seq, 0.55 + 0.4 * overlap)


def is_duplicate(a: str, aliases_a: list[str], b: str, aliases_b: list[str],
                 threshold: float = 0.85) -> bool:
    names = [a, *aliases_a]
    others = [b, *aliases_b]
    return any(name_score(x, y) >= threshold for x in names for y in others)


async def find_duplicate_candidates(nodes: list[dict], threshold: float = 0.85) -> list[tuple[str, str, float]]:
    """Blocking by last-token of normalized name, then pairwise scoring."""
    blocks: dict[str, list[dict]] = {}
    for n in nodes:
        last = normalize(n.get("name", "")).split()
        key = last[-1] if last else ""
        blocks.setdefault(key, []).append(n)
    out: list[tuple[str, str, float]] = []
    for group in blocks.values():
        for i in range(len(group)):
            for j in range(i + 1, len(group)):
                a, b = group[i], group[j]
                s = name_score(a.get("name", ""), b.get("name", ""))
                if s >= threshold or is_duplicate(
                    a.get("name", ""), a.get("aliases", []),
                    b.get("name", ""), b.get("aliases", []), threshold,
                ):
                    out.append((a["id"], b["id"], round(s, 3)))
    return out

File: packages/shared/src/entityResolution.ts
/**
 * Entity resolution: pure, testable duplicate scoring for background merges.
 * Ported from backend/app/services/entity_resolution.py.
 *
 * Scoring mirrors CPython's difflib.SequenceMatcher (gestalt ratio over
 * recursive longest-match blocks; autojunk never triggers on short names),
 * so thresholds behave the same as the Python service.
 */

export function normalize(name: string): string {
  let s = name.normalize("NFKD").replace(/[\u0300-\u036f]/g, "");
  s = s.toLowerCase();
  s = s.replace(/[.\-_,]/g, " ");
  s = s.replace(/\s+/g, " ").trim();
  s = s.replace(/^(mr|mrs|ms|dr|prof)\s+/, "");
  return s;
}

/** Gestalt ratio a.k.a. difflib.SequenceMatcher(None, a, b).ratio(). */
export function gestaltRatio(a: string, b: string): number {
  const m = matchingBlocks(a, b);
  let matched = 0;
  for (const [, , size] of m) matched += size;
  const total = a.length + b.length;
  return total === 0 ? 1 : (2 * matched) / total;
}

type Triple = [number, number, number]; // alo, blo, size

function buildIndex(s: string): Map<string, number[]> {
  const idx = new Map<string, number[]>();
  for (let i = 0; i < s.length; i++) {
    const ch = s[i]!;
    const list = idx.get(ch);
    if (list) list.push(i);
    else idx.set(ch, [i]);
  }
  return idx;
}

function findLongestMatch(
  a: string,
  bIdx: Map<string, number[]>,
  b: string,
  alo: number,
  ahi: number,
  blo: number,
  bhi: number,
): Triple {
  let bestI = alo;
  let bestJ = blo;
  let bestSize = 0;
  // j2len: longest match ending at each b position (difflib algorithm)
  let j2len = new Map<number, number>();
  for (let i = alo; i < ahi; i++) {
    const newJ2len = new Map<number, number>();
    for (const j of bIdx.get(a[i]!) ?? []) {
      if (j < blo) continue;
      if (j >= bhi) break;
      const k = (j2len.get(j - 1) ?? 0) + 1;
      newJ2len.set(j, k);
      if (k > bestSize) {
        bestI = i - k + 1;
        bestJ = j - k + 1;
        bestSize = k;
      }
    }
    j2len = newJ2len;
  }
  // Extend match with non-junk — no junk defined, so extend greedily both ways
  while (bestI > alo && bestJ > blo && a[bestI - 1] === b[bestJ - 1]) {
    bestI--;
    bestJ--;
    bestSize++;
  }
  while (bestI + bestSize < ahi && bestJ + bestSize < bhi && a[bestI + bestSize] === b[bestJ + bestSize]) {
    bestSize++;
  }
  return [bestI, bestJ, bestSize];
}

function matchingBlocks(a: string, b: string): Triple[] {
  const bIdx = buildIndex(b);
  const queue: Array<[number, number, number, number]> = [[0, a.length, 0, b.length]];
  const matches: Triple[] = [];
  while (queue.length > 0) {
    const [alo, ahi, blo, bhi] = queue.pop()!;
    const [i, j, k] = findLongestMatch(a, bIdx, b, alo, ahi, blo, bhi);
    if (k > 0) {
      matches.push([i, j, k]);
      if (alo < i && blo < j) queue.push([alo, i, blo, j]);
      if (i + k < ahi && j + k < bhi) queue.push([i + k, ahi, j + k, bhi]);
    }
  }
  matches.sort((x, y) => x[0] - y[0] || x[1] - y[1]);
  matches.push([a.length, b.length, 0]);
  return matches;
}

function tokenOverlap(ta: Set<string>, tb: Set<string>): number {
  // Single-letter initials count as prefix matches ('n' ~ 'natarajan').
  if (ta.size === 0 || tb.size === 0) return 0;
  const [small, big] = ta.size <= tb.size ? [ta, tb] : [tb, ta];
  let matched = 0;
  for (const tok of small) {
    if (big.has(tok)) matched++;
    else if (tok.length === 1 && [...big].some((b) => b.startsWith(tok))) matched++;
  }
  return matched / Math.max(ta.size, tb.size);
}

export function nameScore(a: string, b: string): number {
  const na = normalize(a);
  const nb = normalize(b);
  if (!na || !nb) return 0;
  if (na === nb) return 1;
  const ta = new Set(na.split(" "));
  const tb = new Set(nb.split(" "));
  const overlap = tokenOverlap(ta, tb);
  const seq = gestaltRatio(na, nb);
  // Last-token (family name) must match for a merge candidate.
  const la = na.split(" ").at(-1);
  const lb = nb.split(" ").at(-1);
  if (la !== lb) return Math.min(seq, 0.5);
  return Math.max(seq, 0.55 + 0.4 * overlap);
}

export function isDuplicate(
  a: string,
  aliasesA: string[],
  b: string,
  aliasesB: string[],
  threshold = 0.85,
): boolean {
  const names = [a, ...aliasesA];
  const others = [b, ...aliasesB];
  return names.some((x) => others.some((y) => nameScore(x, y) >= threshold));
}

export interface CandidateNode {
  id: string;
  name: string;
  aliases?: string[];
}

export async function findDuplicateCandidates(
  nodes: CandidateNode[],
  threshold = 0.85,
): Promise<Array<[string, string, number]>> {
  // Blocking by last-token of normalized name, then pairwise scoring.
  const blocks = new Map<string, CandidateNode[]>();
  for (const n of nodes) {
    const parts = normalize(n.name).split(" ");
    const key = parts.at(-1) ?? "";
    const group = blocks.get(key) ?? [];
    group.push(n);
    blocks.set(key, group);
  }
  const out: Array<[string, string, number]> = [];
  for (const group of blocks.values()) {
    for (let i = 0; i < group.length; i++) {
      for (let j = i + 1; j < group.length; j++) {
        const x = group[i]!;
        const y = group[j]!;
        const s = nameScore(x.name, y.name);
        if (
          s >= threshold ||
          isDuplicate(x.name, x.aliases ?? [], y.name, y.aliases ?? [], threshold)
        ) {
          out.push([x.id, y.id, Math.round(s * 1000) / 1000]);
        }
      }
    }
  }
  return out;
}
